### DISCUSSION TITLE
内存中的数据量是否会影响性能？

### DISCUSSION THREAD

**Discussion by VIVALXH** [apache/kvrocks#2695] [中文]
我们使用时观察到，当有数据不断有写入时，内存不断的上升会带来ops的下降，直到某个机制清理了内存中的数据，ops才逐渐恢复。
请问是什么机制清理了内存中的数据，是否可以通过命令主动控制或者配置？我们尝试过COMPACT命令，修改过rocksdb.block_cache_size、max_write_buffer_number等，但好像并没有效果。

![image](https://github.com/user-attachments/assets/d653878d-e35d-4fe7-acc5-5241edd56df9)

**Comment by git-hulk** [comment]
从以上的截图看 OPS 下降跟 CPU 升高正相关，应该是 Flush/Compaction IO 带来的影响，内存 Cache 对于写入并不会有直接影响。

**Comment by VIVALXH** [comment]
@git-hulk 谢谢，那时的内存下降可能是内存溢出docker重启了，这个表现确实和内存没关系。

在控制内存大小的情况下，我们简化了测试命令并重新测试，应该是lrange的时间复杂度问题。
再请问下，redis lrange是 O(S+N)，但在 kvrocks 代码中，感觉 lrange 的时间复杂度应该是O(N)？start offset会影响吗？
![image](https://github.com/user-attachments/assets/a28fb5a9-4114-4713-874c-36b04b84d253)


测试命令：
redis-benchmark -h 192.168.1.115 -p 6666 -c 30 -n 10000000 eval '
  local key="key:" .. math.random(100000, 200000) 
  local value=string.rep("a", 512) 
  redis.call("lpush",key,value)
  return redis.call("lrange",key,0,10)
' 0

测试结果：
![image](https://github.com/user-attachments/assets/97da2c78-9834-4705-a05c-c976770d3db1)

**Reply by git-hulk** [reply]
> 再请问下，redis lrange是 O(S+N)，但在 kvrocks 代码中，感觉 lrange 的时间复杂度应该是O(N)？start offset会影响吗？

不是 O(N), 一次 RocksDB 的 Seek 操作。
